cpp/src/arrow/dataset/discovery.cc

```diff
@@ -58,57 +62,88 @@ bool StartsWithAnyOf(const std::vector<std::string>& prefixes, const std::string
   return std::any_of(prefixes.cbegin(), prefixes.cend(), matches_prefix);
 }
 
-Result<DataSourceDiscoveryPtr> FileSystemDataSourceDiscovery::Make(
-    fs::FileSystemPtr fs, fs::FileStatsVector files, FileFormatPtr format,
-    FileSystemDiscoveryOptions options) {
-  DCHECK_NE(format, nullptr);
-
-  bool has_prefixes = !options.ignore_prefixes.empty();
-  std::vector<fs::FileStats> filtered;
-  for (const auto& stat : files) {
-    if (stat.IsFile()) {
-      const std::string& path = stat.path();
+Status FileSystemDataSourceDiscovery::Filter(const fs::FileSystemPtr& filesystem,
+                                             const FileFormatPtr& format,
+                                             const FileSystemDiscoveryOptions& options,
+                                             fs::FileStatsVector* files) {
+  for (auto it = files->begin(); it != files->end();) {
+    if (StartsWithAnyOf(options.ignore_prefixes, it->path())) {
+      files->erase(it);
```

Comment:
I'm not a fan of modifying an iterable while iterating. I'd say keep the function but for a single FileStats, and move the exclusion + loop at callsite. Don't modify in place, just create a second array, one more reason for ranges.

It's also bad for performance, if you're deleting most of the elements, it's approaching O(n^2) (n-1 + n-2 + ...) move of elements in a std::vector.

Reply:
Alright.

I would've just used `remove_if` but STL algorithms don't play well with Status return. Handling that with ranges is going to be a challenge